### DISCUSSION TITLE
How does LabGym identify different individuals of the same species?

### DISCUSSION THREAD

**Discussion by prashers** [umyelab/LabGym#92] [Q&A]
Hello, I am interested in trying out LabGym for detecting quails in a flight pen that were recorded from a top-down view (see image below). I have installed the software successfully and produced some images to annotate using Roboflow. 

It is extremely important for my work to know the identities of each individual in my videos (e.g., I need to know whether birdA pecked birdB or whether the peck actually came from birdC). I am wondering how LabGym accomplishes this. 

When annotating images in Roboflow, do I need to have a unique class for each individual so that the detector learns to identify each individual bird? Or does the individual identification happen at a later step?

Thanks!

![GX010911_reencoded_1000](https://github.com/umyelab/LabGym/assets/65912929/02b7b40e-6286-437a-b923-2a5bebd06d2c)

**Comment by yujiahu415** [comment]
Hi,
I noticed that there was a color marker for each bird, right? If the color is visually easy to identify all the time in the video, then you can try annotating each bird as a separate class, like "green bird", "red bird"...In this way, you don't need to worry about ID swap. Otherwise you just use one class "bird" to label all the individuals. During analysis, the IDs (will be shown as numbers) will be automatically assigned for each individual at the first frame, and LabGym will then track the IDs over time. In this case, ID swap might happen. To prevent that, annotate as many frames in which two or more birds collide as possible. If the Detector sees most colliding scenarios during training, ID swap can be significantly avoided. If you want to 100% eliminate ID swap, train a dedicated Detector for each video. When training such kind of Detectors, use the original frame size or its approximation (if the original is not an even number) as inferencing frame size, and use the maximum iteration number (5000).

**Reply by prashers** [reply]
Thank you for the response!

Yes, the birds are individually identifiable by the colors on their backs as well as the unique codes on their backs. 
So to clarify, if I annotate images with a unique class per individual as you said, would the detector learn to identify the individuals based on the colors and/or codes? And this would 100% eliminate ID swaps (that is what I am hoping for)? 
I don't think training a dedicated Detector for each video would be feasible for me. 

One more question: 
I have several shorter video clips for each of my 6 hour trials (videoclip2 starts exactly when videoclip1 ends and so on). So once I get to the stage of feeding new videos into LabGym for automatic behavior detection, will the program keep track of the individuals across these video clips if I import them altogether?

Thanks again!

**Reply by yujiahu415** [reply]
'if I annotate images with a unique class per individual as you said, would the detector learn to identify the individuals based on the colors and/or codes'
--Detector can learn to distinguish color and appearance differences. In your case, only three birds have color marker (green, red, and yellow), others are white. The differences among different codes are very subtle even to human eyes. I'm not sure if Detector can pick up that subtle difference but you can do a quick test: annotate ~50 images and train a Detector (BTW, what is the original frame size?). When export the COCO format dataset of the annotated images, skip the "preprocessing of image" but choose "vertical + horizontal flip" as the augmentation method. After training the Detector, go to "Analysis Behavior", choose "not to categorizer behavior, just track animals", select the trained Detector and run the analysis on the video. LabGym will output a video copy with individual birds annotated with their classes. Check whether the classes match the birds.

If you use class as identity, each class is unique across videoclips, like "green bird" is always that "green bird". So if the Detector accurately distinguishes those identities, LabGym can keep track of the birds across different videoclips. If you use a single class for all birds, the IDs will be assigned as numbers at the first frame of a video, randomly. So the IDs will shuffle at the beginning of each videoclip. To avoid this, you may merge all the clips into one single video.

In your case, since you don't against using markers on birds, my suggestion is to make the marker more visually distinguishable (maybe for the future recordings). That will make the tracking a very easy task. For current videos, you can run a quick test, as suggested above, to see whether the Detector has any chance to succeed. You may let me know how's the test and I can provide feedback whether it can work and how to improve.

**Reply by prashers** [reply]
Each bird does have a different color, but some are more similar to each other - they are distinguishable to a human observer on a large screen. Unfortunately, I will not be collecting any more video recordings, so I will not be able to increase the size of the markers. 

The original frame size is 1440 x 1920.

I annotated just under 50 images with a different class per individual bird based on their color and then trained a detector following your directions. I was running the analysis (just tracking the 6 animals) for some videos, but it seemed to be taking a very long time (about 1 hour for 1000 frames). I had to stop it after a few hours. I have 6 hours of video (shot at 24 fps) per trial, so I was hoping the analysis would not take that much longer than the length of the videos. 

Is this amount of time typical for the analysis step, or can I adjust some settings to increase the speed?
I did not resize the frames when it gave me the option, which may have slowed it down?
Also, I do have a GPU, but I am not sure if it was found and used by LabGym.

**Reply by yujiahu415** [reply]
1. I suspect your GPU hasn’t been detected. The Detectors use PyTorch. You can follow this to check whether your PyTorch can detect your GPU: https://stackoverflow.com/questions/48152674/how-do-i-check-if-pytorch-is-using-the-gpu.
2. You can downsize the video frame size to half, like 720x960, or even 1/4, to make the analysis faster. And also downsize inferencing frame size of the Detector to 720 or even 480.
3. How long is one video? And also, what is the type of the GPU and VRAM? Knowing these information can help me to provide more specific suggestions.
Thanks!

**Reply by prashers** [reply]
Yes, it seems the GPU is not detected. torch.cuda.device_count() returns zero and current_device() returns a message saying 'torch not compiled with CUDA enabled'

Each video clip is about 18 minutes long. 

GPU: NVIDIA GeForce RTX 3060 Ti
VRAM: 8038 MB

**Reply by yujiahu415** [reply]
Try to install PyTorch with CUDA v11.8:

1. Install CUDA toolkit v11.8 (https://developer.nvidia.com/cuda-11-8-0-download-archive?target_os=Windows&target_arch=x86_64) and download CUDNN library that supports 11.x (https://developer.nvidia.com/cudnn-downloads?target_os=Windows) (you need to register a free account to download CUDNN). Place the CUDNN library inside 'C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.8\'. And add cuda/cudnn paths to the environment variables.

2. Install tensorflow v2.10.0 (pip install tensorflow==2.10.0) because after v2.10.0, tensorflow no longer support GPU on native Windows system (but the later versions support Windows subsystem for Linux).

3. Install PyTorch v2.0.1+cu11.8: https://pytorch.org/get-started/previous-versions/. (python3 -m pip install torch==2.0.1 torchvision==0.15.2 torchaudio==2.0.2 --index-url https://download.pytorch.org/whl/cu118)

Reboot your system. And check whether PyTorch can detect your GPU again.

**Reply by prashers** [reply]
I am having some issues getting it to detect the GPU. I think a couple things are making this more complicated for me:

1. I already have a few different versions of the CUDA toolkit on my computer (v11.4, v11.7, v11.8) 
2. I am trying to use labgym within an anaconda virtual environment and I am confused about how to direct it to the correct version of CUDA toolkit

Installing tensorflow and pytorch did not seem to be a problem.

Any suggestions?

**Reply by yujiahu415** [reply]
1. Chech this https://stackoverflow.com/questions/52410484/get-the-cuda-and-cudnn-version-on-windows-with-anaconda-installe to see which version of CUDA toolkit and CUDNN you are using in conda environment.

To direct the CUDA to the correct version, you need to add their path to the environment variables. Check this for how to do so: https://medium.com/geekculture/install-cuda-and-cudnn-on-windows-linux-52d1501a8805. It's an old blog of CUDA v11.0 but the steps are basically the same, just change the versions of CUDA and CUDNN.
Anyway, make sure you not only install CUDA toolkit v11.8 but also download the CUDNN and place the CUDNN library inside 'C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.8\. You may need to replace your the files. And add both CUDA_PATH and CUDNN into your environment variables.

2. Which versions of tensorflow and torch are you using? `import tensorflow as tf` and `import torch` after initiating conda and the type `tf.__version__`, it should return "2.10.0" and type `torch.__version__` it should return "2.0.1+cu118". If not, reinstall them to the correct versions following the "conda installation" in the websites of my previous posts.

Let me know how this goes.

**Reply by prashers** [reply]
I was finally able to get the correct versions of CUDA toolkit (11.8), CUDNN, tensorflow (2.10.0) and pytorch (2.0.1) installed in a fresh anaconda environment running python 3.10.10. 

The GPU is now detected when I follow the directions here: https://stackoverflow.com/questions/48152674/how-do-i-check-if-pytorch-is-using-the-gpu

However, I am now having trouble installing detectron2 in this environment. I used this line from the LabGym user guide: python -m pip install git+https://github.com/facebookresearch/detectron2.git

It returns a lot of text ending with this:
5 errors detected in the compilation of "C:/Users/sanja/AppData/Local/Temp/pip-req-build-6p2q18hc/detectron2/layers/csrc/nms_rotated/nms_rotated_cuda.cu".
      nms_rotated_cuda.cu
      error: command 'C:\\Program Files\\NVIDIA GPU Computing Toolkit\\CUDA\\v11.8\\bin\\nvcc.exe' failed with exit code 1
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for detectron2
  Running setup.py clean for detectron2
Failed to build detectron2
ERROR: Could not build wheels for detectron2, which is required to install pyproject.toml-based projects

Do you have any suggestions to address this?

**Reply by yujiahu415** [reply]
Hi,

1. Upgrade the setuptools, wheel, and pip by `py -m pip install --upgrade pip wheel setuptools`
2. Follow the step 2 in https://labgym.readthedocs.io/en/stable/installation/windows.html to install "Visual Studio C++ Build Tools"
3. Then try to install Detectron2 again by `py -m pip install git+https://github.com/facebookresearch/detectron2.git` to see if you still have the issue.

@rohansatapathy can you also take a look into this installation issue? Thanks!

**Reply by yujiahu415** [reply]
Here are also some useful information from @dr-moo, in which the steps 3, and 6~11 may also be helpful to solve the issue:

> But here is the story related to LabGym :-) It was maddening to install LabGym with Detectron2. And the main issue is Detectron itself. None of the instructions you guys gave me worked. It always failed with the last step when compiling the wheel for Detection. While it is known that Detectron does not play nice with Windows, I finally made it work (CUDA 11.8/cudnn 9.0).
> 
> 1. Virtual conda environment with python=3.10
> 2. Within the environment: run pip install cython, and pip install Cmake
> 3. Add CMake to the Windows Environment PATH variables: It is very important to add CMake to the Environment PATH variables. To do this, download the latest release of CMake from https://perso.uclouvain.be/allan.barrea/opencv/building_tools.html and install it according to the instructions provided on the webpage.
> 4. Install LabGym: pip install labgym
> 5. Install the latest torch: pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118
> 6. From within the intended virtual environment: git clone https://github.com/facebookresearch/detectron2.git
> 7. Find the Detectron2 folder. For me, it got dumped into /Users/NAME/detectron2
> 8. In the detectron2 folder, open setup.py file and make sure you add this line "-DWITH_CUDA",  within the following snippet (right after: "-D__CUDA_NO_HALF2_OPERATORS__"):
> 
> if (torch.cuda.is_available() and ((CUDA_HOME is not None) or is_rocm_pytorch)) or os.getenv( "FORCE_CUDA", "0" ) == "1": extension = CUDAExtension sources += source_cuda
> 
> ```
>     if not is_rocm_pytorch:
>         define_macros += [("WITH_CUDA", None)]
>         extra_compile_args["nvcc"] = [
>             "-O3",
>             "-DCUDA_HAS_FP16=1",
>             "-D__CUDA_NO_HALF_OPERATORS__",
>             "-D__CUDA_NO_HALF_CONVERSIONS__",
>             "-D__CUDA_NO_HALF2_OPERATORS__",
>              "-DWITH_CUDA",
>         ]
>     else:
>         define_macros += [("WITH_HIP", None)]
>         extra_compile_args["nvcc"] = []
> ```
> 
>  10. Save setup.py
> 11. Install Detectron2 with pip install -e detectron2
> 12. Enjoy your LabGym experience!

**Reply by rohansatapathy** [reply]
@yujiahu415 Will do! I’m going to be on vacation until Wednesday, so I’ll take a look at it when I get home on Thursday.

**Reply by prashers** [reply]
OK, I think it worked when I followed all of the directions from @dr-moo in a new conda environment! 
The GPU is still detected by torch as well.
Thank you so much for sharing those instructions and thanks to @dr-moo!

I was going to get back to testing the detector that I had trained (before I ran into the GPU issue, etc.), but I am getting the error below when I select the directory of the trained detector within the analysis module (the step where I specify the method to detect animals).

Traceback (most recent call last):
  File "C:\Users\sanja\miniconda3\envs\labgym4\lib\site-packages\LabGym\gui\analysis\analyze_behaviors.py", line 694, in select_method
    self.path_to_detector = dialog2.GetPaths()
TypeError: DirDialog.GetPaths(): not enough arguments


Hopefully this is an easy fix compared to the installation issues(?)

**Reply by yujiahu415** [reply]
Hi,

This might be a bug in the recent version. @rohansatapathy Can you take a look into this?

The LabGym code has been restructured in v2.3.x but all the functions are the same as v2.2.2, which is a stable version. So an easy fix could be simply downgrading LabGym to its v2.2.2 by `pip install LabGym==2.2.2` in your conda environment.

To run LabGym v2.2.2, you need to first initiate your python by `py` or `python3`, and then type `from LabGym import gui`, and then type `gui.gui()` to initiate LabGym user interface. This is a bit different from v2.3.x

Sorry for this issue and let me know if you still have the same issue in v2.2.2. Thanks!

**Reply by prashers** [reply]
Unfortunately I still get an error at the same step, but it is coming from a different python script:

Traceback (most recent call last):
  File "C:\Users\sanja\miniconda3\envs\labgym4copy\lib\site-packages\LabGym\gui_analyzers.py", line 494, in select_method
    self.path_to_detector=dialog2.GetPaths()
TypeError: DirDialog.GetPaths(): not enough arguments

**Reply by yujiahu415** [reply]
I see. I think the cause of this issue was the new directory to your trained Detector could not be read properly. A quick fix would be: find the "detector" folder in "C:\Users\sanja\miniconda3\envs\labgym4copy\lib\site-packages\LabGym\", then copy the entire folder of your trained Detector into that folder. Then you can find the name of trained Detector will appear in the GUI when you select the detection method. The "detectors" folder stores all the trained detectors by default.

**Reply by prashers** [reply]
This one was definitely a user problem - I was confused about where my trained detector was stored. 
So there is no bug in the software. Sorry about that!

Using the latest version of LabGym, I was able to train a detector and run some video clips through the analysis module to see if it could track individual birds. It was processing 1000 frames in about 30 seconds this time (compared to 1 hour without the GPU and not resizing the image). 
Unfortunately (based on the annotated video) there are many frames where individual birds are not detected at all, even if they are separated from other birds. Sometimes the labels are incorrect as well. There is a digital clock in the frame and it was incorrectly identified as a bird as well.

My original frame size is 1440x1920
I trained the detector on 45 labelled images with an inferencing frame size = 960 and iteration number = 2000.
For the analysis, I entered a resize value = 480 

Is there anything I can do to increase the accuracy of the detector other than training it on more images?

**Reply by yujiahu415** [reply]
Thanks for the update!

Yes, there are many ways to enhance the accuracy of the Detectors.

1. Make sure you select the right steps when exporting your annotated image dataset from Roboflow. 1) Make sure to choose the right format "coco segmentation". 2) SKIP any of the preprocessing steps (like resizing to 640x640) 3) Select some augmentation methods to increase the number and diversity of the dataset. You can definitely select "flipping" (both horizontal and vertical). If the illumination is varying during the video, you can also select "brightness", but if not, no  need to choose this. 

2. There are several settings during training can be optimized. First, increase the iteration number to 5000 and observe the "total loss" during training, ideally it is below 0.3 at the end of training. Second, if you train the Detector at 960 inferencing frame size, ideally you resize the frame size of your video to the same 960 during analysis. But that would make the speed slower. I'm not sure if the slower speed would be a problem for you. If not, I highly recommend this. But if the speed is a problem, you can try to train a Detector with the inferencing frame size==640 and resize video frame to 640 during analysis.

3. If you did 1 and 2 but the accuracy is still low, adding more annotated images. 45 annotated images are on the low side in general, which would be good for a quick test but not for a solid training of an accurate Detector. You can increase the number gradually by adding additional 50 images every time. After adding new annotated images, when exporting them, make sure the augmentation is done on the newly added images as well. If you already have ~200 images before augmentation but the accuracy is still low, which is very rare, let me know. That would indicate something else is wrong.

**Reply by dr-moo** [reply]
I would also like to chime in again, @prashers! What is your setting? Are we talking broilers (my first assumption when I hear birds)? Research box? Any enrichment? Walkway with single individual? Several birds, and if that, how many? Commercial setup with many birds at the same time? And how is the camera placed - angled, top-down? Height? All these details matter and could steer the number of initial frames for annotation.

For example, when we had some projects with broilers and used segmentation for detection, we tried the commercial setup with several thousand birds in a frame and an enriched pen with 15-20 individuals. For the research pen, we had roughly 260 frames with 5000+ birds manually annotated, and for the commercial farm, we had 15 high-resolution frames with 7000 birds annotated across the frames. It all depends on your initial research question and scene variability. But the first iteration of algorithm development is ALWAYS time-consuming, and it is worth it.

When selecting frames for annotation, it could also be helpful to think in terms of dynamic scenes and pick variable sequences with as much change as possible but still close to the desired scene/event type you want to analyse, thus somewhat minimising the manual labour.

**Reply by yujiahu415** [reply]
Great inputs! @dr-moo 
I guess the number of birds in @prashers 's scenarios would not be too many. But the identity of each bird is important. It's a good point that when selecting frames, you need to consider the balance between making the training dataset as diverse as possible and making it still close to the scene in analysis. A more extreme case is that you just annotate the frames in the videos to analyze to minimize the detection difficulty--after all you're not testing the generalizability of the Detectors but just want to make it as accurate as possible to enable analysis.

**Reply by prashers** [reply]
Thank you both for the input/suggestions! 

@dr-moo , I have 6 birds in an indoor flight pen recorded from a top-down view (see picture in my first post in this thread). I really appreciate you giving me a reference point for how many frames it might take for me to get reliable results!  

**Updates:**
Simply making sure I resized the frames to match the inferencing frame size (960) improved the tracking dramatically! It is still mixing up bird identities quite a bit (especially between birds wearing the white, black and blue markers), but it seems promising otherwise. I will be trying the other suggestions you mentioned to see if things improve. I will definitely be annotating more images - 45 was just for the initial test.

I would like the analysis to run a bit faster than it did for a frame size of 960, so I trained another detector with an inferencing frame size of 640 and 5000 iterations (the total_loss got down to 0.1523). **I tried to analyze a video resized to 640 and it was running as normal, until after 15000 frames processed it returned this error**:

`C:\Users\sanja\miniconda3\envs\labgym4\lib\site-packages\LabGym\analyzebehaviorsdetector.py:857: RuntimeWarning: invalid value encountered in divide
  np.sum(np.logical_and(masks[:, None], masks), axis=(2, 3))
C:\Users\sanja\miniconda3\envs\labgym4\lib\site-packages\LabGym\analyzebehaviorsdetector.py:857: RuntimeWarning: invalid value encountered in divide
  np.sum(np.logical_and(masks[:, None], masks), axis=(2, 3))
Traceback (most recent call last):
  File "C:\Users\sanja\miniconda3\envs\labgym4\lib\site-packages\LabGym\gui\analysis\analyze_behaviors.py", line 1372, in analyze_behaviors
    AAD.acquire_information(
  File "C:\Users\sanja\miniconda3\envs\labgym4\lib\site-packages\LabGym\analyzebehaviorsdetector.py", line 1420, in acquire_information
    self.detect_track_individuals(
  File "C:\Users\sanja\miniconda3\envs\labgym4\lib\site-packages\LabGym\analyzebehaviorsdetector.py", line 927, in detect_track_individuals
    sorted(cnts, key=cv2.contourArea, reverse=True)[0]
IndexError: list index out of range`

**Reply by yujiahu415** [reply]
This error is simply because a detection failure, in which the contour cannot be closed and its area cannot be calculated, causing an opencv error. You can simply increase the frame size or train a Detector with more annotated images.

**Reply by yujiahu415** [reply]
The ID mix is possibly because the white, black, and blue markers are visually more difficult to distinguish. I guess the Detector had higher accuracy in distinguishing IDs with green and red markers?

**Comment by dr-moo** [comment]
Sorry, @prashers, i missed the picture initially, but i understand what you mean now. As @yujiahu415 points out, some markers will always be more difficult to distinguish. It does not mean it is impossible to solve, but robust tracking will require additional steps. You could always (and it is usually the best idea) include more annotated frames, where you make sure you have as many occluding birds as possible, as well as birds with different markers) to improve the overall detector accuracy. But this is only good until certain point, and as animal scientist, you might need to have a clear cut off point, where time investment for annotation becomes almost as extensive as just asking a student to look at the video manually (or if you do it yourself). 

Another strategy is to accept the potential ID switches and work with log files, where you could stitch tracelets together by filtering out false IDs in between frames. This is still manual work but not as extensive as fully manual observations. 

And path number three, if the continuous tracking is an absolute priority - Yolo (any version from 4 and up) in combination with SORT/DeepSORT work very well in setups like yours, but then you don't have user-friendly interface and comfort of LabGym.

**Reply by yujiahu415** [reply]
Thank you again, @dr-moo, for the insightful comments!!
@prashers, some experience / tips:
1. As @dr-moo mentioned, you only need to select the frames in which colliding occurs for annotation. In other word, you don't need to annotate any frames in which all birds are well separated.
2. If the green and red markers work well for distinguishing the two birds, keep them as separate classes, like "green bird" and "red bird". For another markers that didn't work well, you can have only one class, like "other bird", and annotate more frames. In my hand, for a video of 5 similar-looking, interacting flies (640X640, 10-minute duration, 25 fps), if I annotate ~100 frames in which collision / occlusion occurs, with only one class, "fly", and trained a Detector for this video at 640 inferencing frame size with 5,000 iterations, the identity switches are typically less than 5. If I annotate ~300 frames in which collision / occlusion occurs, the identity switch can be 0.

**Reply by prashers** [reply]
Thanks for the additional suggestions. 

I took a step back to think about whether the time investment would be worth it for my project. I realized I am not sure if LabGym would actually give me the data I am hoping for. I am hoping to get information on the frequency of social interactions between particular birds to understand how their relationships change over time. For example, I want to know how many times blue pecked green in a trial (not just the number of times blue pecked **any** bird in general), or how many times yellow groomed red (not just the number of times yellow groomed **any** other bird). 

Would LabGym give me information about the actor and receiver of each social interaction, or is it only able to tell me how many times the focal bird exhibited each behavior I define?

**Reply by yujiahu415** [reply]
LabGym currently can only tell you at a time point which bird "is pecking" (any other bird) and which bird "is being pecked" (by any other bird). So if there are multiple "pecking" and "being pecked" pairs occur at the same time point, you need additional posthoc analysis (e.g., calculate the distance between all possible "pecking" bird and "being pecked" bird and sort the shortest distance) to link the two birds that have the relationship.

**Reply by prashers** [reply]
I understand. After some more thought, I think it might be too difficult/time consuming to get LabGym working for my purposes. I really appreciate all of your advice! I am sure it could prove to be useful for myself or my colleagues in the future.

Out of curiosity, is there anything in development to associate each detected social behavior with a particular actor and receiver (especially when there may be multiple behaviors happening simultaneously)? I think that would be a great feature for people studying social behavior like myself!

**Reply by yujiahu415** [reply]
No problem! And thank you for your effort! We will implement new features to address the "multiple pairs of particular actor and receiver" problem in future versions. As I mentioned in previous responses, this is doable because LabGym can already detect "actor" and "receiver", we only need to introduce an additional step of sorting out the distance between all possible actors and receivers to match the particular pairs.
